packages/nimble-components/src/wafer-map/specs/features/change-highlight.md

```diff
@@ -0,0 +1,59 @@
+# Nimble Wafer Map Die Highlighting
+
+## Problem Statement
+
+The wafer map receives an array of strings as the values that will be highlighted.
+
+```
+@observable public highlightedValues: string[] = [];
+```
+
+In the following case the highlighted values are 2 and 3.
+
+![Highlighted Values](resources/highlighted-values.png)
+
+The current approach is limited. We can either highlight all the dies having 2s and 3s or neither of them. The customer needs exceeds this approach. The customer may see an anomaly on the dies having DieX: 15 and Soft Bin: 2 and wishes to highlight all these values, but the current code is not capable of this.
+
+## Links To Relevant Work Items and Reference Material
+
+N\A
+
+## Implementation / Design
```

Comment:
I thought of a couple other potential APIs and I'm curious for your opinions on whether they would be preferable:

1. remove `HighlightMode` and instead highlight the union of dies from `highlightedValues` and `isHighlighted`. That way:
    - `HighlightMode.Nothing` option could be achieved by leaving both `highlightedValues` and `isHighlighted` unset
    -  `HighlightMode.Value` option could be achieved by leaving `isHighlighted` unset
    - `HighlightMode.Coordinates` option could be achieved by leaving `highlightedValues` unset
2. Remove `highlightedValues` entirely and instead require clients to update `isHighlighted` on every `WaferMapDie` with a particular value

From an API design perspective I think both options are preferable to the current proposal because the API surface area is smaller so there are fewer cases to test for (with 2 being even simpler than 1, obviously). But I don't have a sense for the impact on clients or the complexity/performance implications.

Reply:
Good ideas, I will go with the first one

Reply:
It kinda feels like we are conflating data with presentation a bit in the api / conflating incidental view state with the data. Feels similar to the selection state of rows on the table.

Some thoughts:
- I think the main problem with highlightedValues is that it can't address independent wafers uniquely
- Updating the data is an expensive operation. Lets say you wanted to update the highlighted state interactively with mouse moves. We shouldn't need to reset the whole data set, just the highlighted state.
- Maybe we want additional incidental state in the future like `selected` or `loading` states for specific wafers

Maybe we should do something more like the table:
- Include an `id?: string` in `WaferMapDie`
- Keep configuration of the highlighted wafers separate, i.e. `highlightedDies: string[] = []`

Concretely, the alternate proposals are:
1. Add an optional `id` to `WaferMapDie`. Validate that all ids are unique or all are are absent. Add `highlightedDies` which takes an array of `id`s. If any ids in `highlightedDies` don't match the data in `dies` just ignore them.
2. Just use the index in `dies` as the identifier for a wafer and don't add an explicit `id`  field. `highlightedDies` just takes an array of indexes.

In all of these proposals I think we should get rid of `highlightedValues`. I don't think there is much value in having both `highlightedValues` and `highlightedDies:[]` / `[{isHighLighted}]`.

Reply:
The concerns you raised make sense; we would not want to update the data just for the highlight change.

But I also have a concern regarding the id/index approach. The wafer map use-case, differently from the table example you gave, will not have just a small number of dies highlighted. In some cases, we would have most of the dies selected based on the user selecting some die traits. I worry that this will impact the performance of the renderer when it must find the specific id/indexes in the provided list. Do you think this is a valid concern?

As an alternative we are thinking of using secondary values in the metadata which can be used for highlighting along with a secondary highlighting list. 

```
dies: [
   {
       x: 1,
        y: 1,
        value: 12,
        metadata: {
            secondaryValue: 2
        }
    },
    ...
]

secondaryHighlightList: [{secondaryValue: 2}, {secondaryValue: 3} ...]
```
This can cover the scenario you presented if the user will use ids for secondary values and a small list of ids in the secondary highlight list and it can also cover the scenario when we have different die traits from the displayed value that we may want to highlight.

We can also make this list the only highlight list, and when the metadata is absent it will use the actual values.

Reply:
What if instead of unique ids we did tags?

```
dies: [{
  x:1,
  y:1,
  value: 12,
  tags: ['a', 'b']
}]
highlightedDies: ['a']
```

Then you have a lot of flexibility from the client. You could create a unique tag per wafer if you wanted and/or create groups of wafers by giving them the same tag. Kinda similar to CSS classes from that perspective.

Reply:
It's a great ideea, having the highlight clearly defined to tags makes it less confusing and we leve the whole highlight scheme to the client, so they can make it suit their performance needs